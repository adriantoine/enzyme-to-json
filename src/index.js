import omit from 'lodash.omit';
import pickBy from 'lodash.pickby';
import compact from 'lodash.compact';
import get from 'lodash.get';

export function shallowToJson(wrapper) {
    const type = wrapper.type();
    if (!type) {
        return wrapper.node;
    }

    const children = compact(wrapper.children().map(c => shallowToJson(c)));
    const props = pickBy(wrapper.props(), val => val !== undefined);

    const json = {
        type: wrapper.name(),
        props: omit(props, 'children'),
        children: (children.length) ? children : null,
        $$typeof: Symbol.for('react.test.json'),
    };

    // If the type of element is not a function, it means that is a DOM element
    if (typeof type !== 'function') {
        return json;
    }

    // We need to get the rendered component in Enzyme internals
    // because it won't be returned by `.children()`
    const element = get(wrapper, 'node._reactInternalInstance._renderedComponent._currentElement');
    if(!element) {
        return json;
    }

    const jsonChildren = json.children;
    json.children = [{
        type: element.type,
        props: omit(element.props, 'children'),
        children: jsonChildren,
        $$typeof: Symbol.for('react.test.json'),
    }];

    return json;
}

export {shallowToJson as mountToJson};

const renderChildToJson = child => {
    if(!child) return;

    if(child.type === 'tag') {
        return {
            type: child.name,
            props: child.attribs,
            children: child.children && child.children.length ? compact(child.children.map(renderChildToJson)) : null,
            $$typeof: Symbol.for('react.test.json'),
        };
    } else if(child.type === 'text') {
        return child.data;
    }
};
export const renderToJson = wrapper =>
    renderChildToJson(wrapper.children()[0]);
