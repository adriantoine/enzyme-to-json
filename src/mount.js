import compact from 'lodash.compact';
import omitBy from 'lodash.omitby';
import values from 'object-values';
import {isDOMComponent, isElement} from 'enzyme/build/react-compat';
import {internalInstance, propsOfNode} from 'enzyme/build/Utils';
import {typeName} from 'enzyme/build/Debug';
import {childrenOfNode} from 'enzyme/build/ShallowTraversal';

function instToJson(inst) {
    if (typeof inst === 'string' || typeof inst === 'number') return inst;
    if (!inst) return '';

    if (!inst.getPublicInstance) {
        const internal = internalInstance(inst);
        return instToJson(internal);
    }
    const publicInst = inst.getPublicInstance();

    if (typeof publicInst === 'string' || typeof publicInst === 'number') return publicInst;
    if (!publicInst && !inst._renderedComponent) return '';

    const currentElement = inst._currentElement;
    const type = typeName(currentElement);
    const props = omitBy(propsOfNode(currentElement), (val, key) => key === 'children' || val === undefined);
    const children = [];
    if (isDOMComponent(publicInst)) {
        const renderedChildren = inst._renderedChildren;
        if (!renderedChildren) {
            children.push(...childrenOfNode(currentElement));
        } else {
            children.push(...values(renderedChildren));
        }
    } else if (
        isElement(currentElement) &&
        typeof currentElement.type === 'function'
    ) {
        children.push(inst._renderedComponent);
    }

    const childrenArray = compact(children.map(n => instToJson(n)));

    return {
        type,
        props,
        children: childrenArray.length ? childrenArray : null,
        $$typeof: Symbol.for('react.test.json'),
    };
}

export function mountToJson(wrapper) {
    return instToJson(wrapper.node);
}
