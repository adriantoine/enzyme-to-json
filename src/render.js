import compact from 'lodash.compact';

const renderChildToJson = child => {
    if(!child) return;

    if(child.type === 'tag') {
        return {
            type: child.name,
            props: child.attribs,
            children: child.children && child.children.length ? compact(child.children.map(renderChildToJson)) : null,
            $$typeof: Symbol.for('react.test.json'),
        };
    } else if(child.type === 'text') {
        return child.data;
    }
};

export default wrapper => renderChildToJson(wrapper.children()[0]);
