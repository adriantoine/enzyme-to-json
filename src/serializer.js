import ShallowWrapper from 'enzyme/build/ShallowWrapper';
import ReactWrapper from 'enzyme/build/ReactWrapper';

import toJson from './';

module.exports = {
    test(wrapper) {
        return wrapper instanceof ShallowWrapper || wrapper instanceof ReactWrapper || wrapper.cheerio;
    },
    print(wrapper) {
        return toJson(wrapper);
    },
};
