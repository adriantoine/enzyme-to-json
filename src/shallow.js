import compact from 'lodash.compact';
import isPlainObject from 'lodash.isplainobject';
import omit from 'lodash.omit';
import entries from 'object.entries';
import {propsOfNode} from 'enzyme/build/Utils';
import {typeName} from 'enzyme/build/Debug';
import {childrenOfNode} from 'enzyme/build/ShallowTraversal';
import {isElement} from 'enzyme/build/react-compat';

function nodeToJson(node) {
    if(!node) return node;

    if (typeof node === 'string' || typeof node === 'number') {
        return node;
    }

    if (!isElement(node)) {
        if (Array.isArray(node)) {
            return node.map(nodeToJson);
        }

        if (isPlainObject(node)) {
            return entries(node).reduce((obj, [key, val]) => {
                obj[key] = nodeToJson(val);
                return obj;
            }, {});
        }

        return node;
    }

    const children = compact(childrenOfNode(node).map(n => nodeToJson(n)));
    const type = typeName(node);
    const props = omit(propsOfNode(node), 'children');

    return {
        type,
        props: nodeToJson(props),
        children: children.length ? children : null,
        $$typeof: Symbol.for('react.test.json'),
    };
}

export function shallowToJson(wrapper) {
    return nodeToJson(wrapper.node);
}
