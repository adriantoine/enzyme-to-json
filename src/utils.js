import ShallowWrapper from 'enzyme/build/ShallowWrapper';
import ReactWrapper from 'enzyme/build/ReactWrapper';

const SHALLOW_WRAPPER_NAME = ShallowWrapper.name;
const REACT_WRAPPER_NAME = ReactWrapper.name;

export const isShallowWrapper = wrapper => wrapper && wrapper.constructor && wrapper.constructor.name === SHALLOW_WRAPPER_NAME;
export const isReactWrapper = wrapper => wrapper && wrapper.constructor && wrapper.constructor.name === REACT_WRAPPER_NAME;
export const isCheerioWrapper = wrapper => wrapper.cheerio;

export const isEnzymeWrapper = wrapper => isShallowWrapper(wrapper) || isReactWrapper(wrapper) || isCheerioWrapper(wrapper);
