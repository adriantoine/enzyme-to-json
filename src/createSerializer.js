import {isEnzymeWrapper} from './utils';
import toJson from './';

export default function createSerializer(options) {
  return {
    test(wrapper) {
      return isEnzymeWrapper(wrapper);
    },
    print(wrapper, serializer) {
      return serializer(toJson(wrapper, options));
    },
  };
}
