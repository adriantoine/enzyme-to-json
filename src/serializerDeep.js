import {isEnzymeWrapper} from './utils';
import toJson from './';

export function test(wrapper) {
  return isEnzymeWrapper(wrapper);
}

export function print(wrapper, serializer) {
  return serializer(toJson(wrapper, {mode: 'deep'}));
}
