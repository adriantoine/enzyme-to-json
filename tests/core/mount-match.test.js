/* eslint-env jest */

import React from 'react';
import {mount} from 'enzyme';
import {mountToDeepJson, mountToShallowJson} from '../../src';
import {elementToObject, Table, TobeList} from './fixtures/match-object';

describe('Table', () => {
  // Mock child components to omit irrelevant props for shallow match,
  // especially event handlers so not to be redundant with interaction tests.
  function TableHead() {}
  function TableRow() {}
  function Button() {}
  function addRow() {}
  function deleteRow() {}

  // suggestions in a game of Cluedo
  const fields = [
    {key: 'who', label: 'qui'},
    {key: 'where', label: 'où'},
    {key: 'which', label: 'lequel'},
  ];
  const records = [
    {id: 3, who: 'Plum', where: 'study', which: 'spanner'},
    {id: 2, who: 'Peacock', where: 'library', which: 'rope'},
    {id: 1, who: 'Green', where: 'conservatory', which: 'pipe'},
  ];

  describe('matches class components to maximum depth', () => {
    const $it = mount(
      <Table
        addRow={addRow}
        deleteRow={deleteRow}
        fields={fields}
        records={records}
      />,
    );

    test('at thead', () => {
      const received = mountToDeepJson($it.find('thead'));
      expect(received).toMatchSnapshot();
      expect(received).toMatchObject(
        elementToObject(
          <thead>
            <tr>
              <th>
                <button title="add row">+</button>
              </th>
              <th>{fields[0].label}</th>
              <th>{fields[1].label}</th>
              <th>{fields[2].label}</th>
            </tr>
          </thead>,
        ),
      );
    });

    test('at tr', () => {
      // Be careful. Don’t duplicate the application code.
      const i = 1;
      const received = mountToDeepJson($it.find('tbody tr').at(i));
      expect(received).toMatchSnapshot();
      expect(received).toMatchObject(
        elementToObject(
          <tr>
            <td key="">
              <button title="delete row">-</button>
            </td>
            {fields.map(({key}) => <td key={key}>{records[i][key]}</td>)}
          </tr>,
        ),
      );
    });
  });

  describe('matches class components to minimum depth', () => {
    const $it = mount(
      <Table
        addRow={addRow}
        deleteRow={deleteRow}
        fields={fields}
        records={records}
      />,
    );

    test('at table', () => {
      const received = mountToShallowJson($it.find('table'));
      expect(received).toMatchSnapshot();
      expect(received).toMatchObject(
        elementToObject(
          <table>
            <TableHead fields={fields} />
            <tbody>
              <TableRow fields={fields} record={records[0]} />
              <TableRow fields={fields} record={records[1]} />
              <TableRow fields={fields} record={records[2]} />
            </tbody>
          </table>,
        ),
      );
    });

    test('at thead', () => {
      const received = mountToShallowJson($it.find('thead'));
      expect(received).toMatchSnapshot();
      expect(received).toMatchObject(
        elementToObject(
          <thead>
            <tr>
              <th>
                <Button title="add row" text="+" />
              </th>
              <th>{fields[0].label}</th>
              <th>{fields[1].label}</th>
              <th>{fields[2].label}</th>
            </tr>
          </thead>,
        ),
      );
    });

    test('at tbody', () => {
      const received = mountToShallowJson($it.find('tbody'));
      expect(received).toMatchSnapshot();
      expect(received).toMatchObject(
        elementToObject(
          <tbody>
            <TableRow fields={fields} record={records[0]} />
            <TableRow fields={fields} record={records[1]} />
            <TableRow fields={fields} record={records[2]} />
          </tbody>,
        ),
      );
    });

    test('at tr', () => {
      // Be careful. Don’t duplicate the application code.
      const i = 1;
      const received = mountToShallowJson($it.find('tbody tr').at(i));
      expect(received).toMatchSnapshot();
      expect(received).toMatchObject(
        elementToObject(
          <tr>
            <td key="">
              <Button text="-" title="delete row" />
            </td>
            {fields.map(({key}) => <td key={key}>{records[i][key]}</td>)}
          </tr>,
        ),
      );
    });
  });
});

describe('TobeList', () => {
  function TobeItem() {} // mock, and provide only relevant props

  // statements which are to be, or not to be :)
  const items = [
    {
      id: 0,
      is: false,
      text: 'testing is painless',
    },
    {
      id: 1,
      is: true,
      text: 'more is less',
    },
    {
      id: 2,
      is: true,
      text: 'less is more',
    },
  ];

  it('matches functional components to maximum depth', () => {
    const received = mountToDeepJson(mount(<TobeList items={items} />));
    const expected = elementToObject(
      <ul>
        <li style={{textDecoration: 'line-through'}}>testing is painless</li>
        <li style={{textDecoration: 'none'}}>more is less</li>
        <li style={{textDecoration: 'none'}}>less is more</li>
      </ul>,
    );

    expect(received).toMatchObject(expected);

    // Callback props usually prevent toEqual assertion.
    expect(received).toEqual(expected);
  });

  it('matches functional components to minimum depth at a descendant', () => {
    const received = mountToShallowJson(
      mount(<TobeList items={items} />).find('ul'),
      {noKey: true},
    );
    const expected = elementToObject(
      <ul>
        <TobeItem text="testing is painless" is={false} />
        <TobeItem text="more is less" is />
        <TobeItem text="less is more" is />
      </ul>,
    );

    expect(received).toMatchObject(expected);

    // Callback props usually prevent toEqual assertion.
    expect(received).toEqual(expected);
  });

  it('matches functional components to either depth at a descendant', () => {
    const element = <TobeList items={items} />;
    const receivedDeep = mountToDeepJson(
      mount(element)
        .find('li')
        .at(1),
    );
    const receivedShallow = mountToShallowJson(
      mount(element)
        .find('li')
        .at(1),
    );
    const expected = elementToObject(
      <li style={{textDecoration: 'none'}}>more is less</li>,
    );

    // When all descendants are DOM nodes, deep and shallow are equivalent.
    expect(receivedDeep).toMatchObject(expected);
    expect(receivedShallow).toMatchObject(expected);

    // Callback props usually prevent toEqual assertion.
    expect(receivedDeep).toEqual(expected);
    expect(receivedShallow).toEqual(expected);
  });
});
