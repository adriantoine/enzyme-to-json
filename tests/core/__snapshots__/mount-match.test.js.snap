// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table matches class components to maximum depth at thead 1`] = `
<thead>
  <tr>
    <th
      key=""
    >
      <button
        onClick={[Function]}
        title="add row"
      >
        +
      </button>
    </th>
    <th
      key="who"
      scope="col"
    >
      qui
    </th>
    <th
      key="where"
      scope="col"
    >
      où
    </th>
    <th
      key="which"
      scope="col"
    >
      lequel
    </th>
  </tr>
</thead>
`;

exports[`Table matches class components to maximum depth at tr 1`] = `
<tr>
  <td
    key=""
  >
    <button
      onClick={[Function]}
      title="delete row"
    >
      -
    </button>
  </td>
  <td
    key="who"
  >
    Peacock
  </td>
  <td
    key="where"
  >
    library
  </td>
  <td
    key="which"
  >
    rope
  </td>
</tr>
`;

exports[`Table matches class components to minimum depth at table 1`] = `
<table>
  <TableHead
    addRow={[Function]}
    fields={
      Array [
        Object {
          "key": "who",
          "label": "qui",
        },
        Object {
          "key": "where",
          "label": "où",
        },
        Object {
          "key": "which",
          "label": "lequel",
        },
      ]
    }
  />
  <tbody>
    <TableRow
      deleteRow={[Function]}
      fields={
        Array [
          Object {
            "key": "who",
            "label": "qui",
          },
          Object {
            "key": "where",
            "label": "où",
          },
          Object {
            "key": "which",
            "label": "lequel",
          },
        ]
      }
      key="3"
      record={
        Object {
          "id": 3,
          "where": "study",
          "which": "spanner",
          "who": "Plum",
        }
      }
    />
    <TableRow
      deleteRow={[Function]}
      fields={
        Array [
          Object {
            "key": "who",
            "label": "qui",
          },
          Object {
            "key": "where",
            "label": "où",
          },
          Object {
            "key": "which",
            "label": "lequel",
          },
        ]
      }
      key="2"
      record={
        Object {
          "id": 2,
          "where": "library",
          "which": "rope",
          "who": "Peacock",
        }
      }
    />
    <TableRow
      deleteRow={[Function]}
      fields={
        Array [
          Object {
            "key": "who",
            "label": "qui",
          },
          Object {
            "key": "where",
            "label": "où",
          },
          Object {
            "key": "which",
            "label": "lequel",
          },
        ]
      }
      key="1"
      record={
        Object {
          "id": 1,
          "where": "conservatory",
          "which": "pipe",
          "who": "Green",
        }
      }
    />
  </tbody>
</table>
`;

exports[`Table matches class components to minimum depth at tbody 1`] = `
<tbody>
  <TableRow
    deleteRow={[Function]}
    fields={
      Array [
        Object {
          "key": "who",
          "label": "qui",
        },
        Object {
          "key": "where",
          "label": "où",
        },
        Object {
          "key": "which",
          "label": "lequel",
        },
      ]
    }
    key="3"
    record={
      Object {
        "id": 3,
        "where": "study",
        "which": "spanner",
        "who": "Plum",
      }
    }
  />
  <TableRow
    deleteRow={[Function]}
    fields={
      Array [
        Object {
          "key": "who",
          "label": "qui",
        },
        Object {
          "key": "where",
          "label": "où",
        },
        Object {
          "key": "which",
          "label": "lequel",
        },
      ]
    }
    key="2"
    record={
      Object {
        "id": 2,
        "where": "library",
        "which": "rope",
        "who": "Peacock",
      }
    }
  />
  <TableRow
    deleteRow={[Function]}
    fields={
      Array [
        Object {
          "key": "who",
          "label": "qui",
        },
        Object {
          "key": "where",
          "label": "où",
        },
        Object {
          "key": "which",
          "label": "lequel",
        },
      ]
    }
    key="1"
    record={
      Object {
        "id": 1,
        "where": "conservatory",
        "which": "pipe",
        "who": "Green",
      }
    }
  />
</tbody>
`;

exports[`Table matches class components to minimum depth at thead 1`] = `
<thead>
  <tr>
    <th
      key=""
    >
      <Button
        onClick={[Function]}
        text="+"
        title="add row"
      />
    </th>
    <th
      key="who"
      scope="col"
    >
      qui
    </th>
    <th
      key="where"
      scope="col"
    >
      où
    </th>
    <th
      key="which"
      scope="col"
    >
      lequel
    </th>
  </tr>
</thead>
`;

exports[`Table matches class components to minimum depth at tr 1`] = `
<tr>
  <td
    key=""
  >
    <Button
      onClick={[Function]}
      text="-"
      title="delete row"
    />
  </td>
  <td
    key="who"
  >
    Peacock
  </td>
  <td
    key="where"
  >
    library
  </td>
  <td
    key="which"
  >
    rope
  </td>
</tr>
`;
